docs/architecture/ADR-007-bad-debt.md

```diff
@@ -0,0 +1,92 @@
+
+Also note that the fist decision on the condition for bad debt repayment eligibility changes the effectiveness of these options:
+
+- A condition of `Collateral Value == 0` favors checks during `LiquidateBorrow`. because liquidation is the only action on undercollateralized accounts that is allowed to reduce collateral to zero.
+- A condition of `Collateral Value < Borrow Value` mandates periodic checks, as such an inequality can result not only from liquidation, but also from interest accrual and/or asset price fluctuations.
+
+## Decision
+
+The simplest way to implement bad debt repayment is as follows:
+
+At the end of `LiquidateBorrow`, which is triggered by `MsgLiquidate`, the `borrowerAddress` in question is checked for `Collateral Value == 0`. If true, a separate function `RepayBadDebt(borrowerAddress)` is called.
+
+The function `RepayBadDebt(borrowerAddress)` immediately repays any remaining borrows on the address in question, or repays the maximum available amount for any borrowed asset denomination where the borrowed amount exceeds available reserves.
+
+### Messages, Events, and Logs
+
+No new message types are required for bad debt repayment, as it is automatic.
+
+The function `RepayBadDebt` must create event `EventRepayBadDebt` and logs for each borrowed asset it repays, recording address, denom, and amount.
+
+An additional event `ReservesExhausted` could be created to monitor for that specific edge case. It should record the `denom`, `borrowerAddress`, and remaining `BorrowedAmount` for any borrow that `RepayBadDebt` fails to repay.
+
+### Edge Cases
+
+There are two edge cases that would allow bad debt to slip past detection or automatic repayment:
+
+- Nonzero collateral: If liquidators do not reduce borrower collateral to zero, even leaving a miniscule amount, bad debt will not be detected at the end of `LiquidateBorrow`.
+
+This could be solved by any Liquidator - but if undercollateralized borrowers are being left for long periods without being completely liquidated, then a bot could be set up to finish them, even if that bot is controlled by Umee.
+
+- Reserve exhaustion: If reserves are insufficient to fully repay bad debt at the moment `RepayBadDebt` is called, the remaining debt will not be targeted again (as with zero collateral, no `LiquidateBorrow` will occur int he future).
+
+An additional trick could be used in the case of reserve exhaustion: If transfer of uTokens is permitted, sending a miniscule amount of uTokens to an address with zero collateral would make it eligible for liquidation again, whereupon its bad debt would be detected and repaid.
+
+If reserve exhaustion is deemed sufficiently rare, then it might be enough to have this trick available should the need for it arise.
+Otherwise, a very infrequent periodic check could be created to detect address the case automatically, or a message type that could be manually used to flag bad debt for repayment.
```

Comment:
It doesn't seem to me that the "trick" to address the 2nd edge-case is sufficient. It just seems way too inferior and hacky.

I would opt to use some sort of automatic system. Perhaps infrequent periodic checks during BeginBlock or EndBlock.

Reply:
See updates - solution proposed and review re-requested.